opentelemetry-stackdriver/tests/generate.rs

```diff
@@ -0,0 +1,234 @@
+        for i in 0..components {
+            let (prev, cur) = (previous.get(i), parent.get(i));
+            if prev == cur && close == 0 && open.is_empty() {
+                continue;
+            }
+
+            match (prev, cur) {
+                (Some(_), Some(new)) => {
+                    close += 1;
+                    open.push(new);
+                }
+                (Some(_), None) => close += 1,
+                (None, Some(new)) => open.push(new),
+                (None, None) => unreachable!(),
+            }
+        }
+
+        // Close modules.
+
+        let closed = close > 0;
+        while close > 0 {
+            for _ in 0..((level - 1) * 4) {
+                root.push(' ');
+            }
+            root.push_str("}\n");
+            close -= 1;
+            level -= 1;
+        }
+
+        if closed {
+            root.push('\n');
+        }
+
+        // Open modules.
+
+        let mut opened = false;
+        for component in &open {
+            if !opened && !closed {
+                root.push('\n');
+                opened = true;
+            }
+
+            for _ in 0..(level * 4) {
+                root.push(' ');
+            }
+
+            root.push_str("pub mod ");
+            root.push_str(component);
+            root.push_str(" {\n");
+            level += 1;
+        }
+
+        // Write a module declaration for this actual module.
+
+        for _ in 0..(level * 4) {
+            root.push(' ');
+        }
+        root.push_str("pub mod ");
+        root.push_str(module.last().unwrap());
+        root.push_str(";\n");
+        previous = parent;
+    }
+
+    fs::write(source_dir.join("mod.rs"), root).unwrap();
+
+    // Move on to actually comparing the old and new versions.
+
+    let versions = [SOURCE_DIR, GENERATED_TEMP]
+        .iter()
+        .map(|path| {
+            let mut files = HashMap::new();
+            for entry in WalkDir::new(path) {
+                let entry = match entry {
+                    Ok(e) => e,
+                    Err(_) => continue,
+                };
+
+                let is_file = entry.file_type().is_file();
+                let rs = entry.path().extension() == Some(OsStr::new("rs"));
+                if !is_file || !rs {
+                    continue;
+                }
+
+                let file = entry.path();
+                let name = file.strip_prefix(path).unwrap();
+                files.insert(name.to_owned(), fs::read_to_string(file).unwrap());
+            }
+
+            files
+        })
+        .collect::<Vec<_>>();
+
+    // Compare the old version and new version and fail the test if they're different.
+
+    let mut keys = versions[0].keys().collect::<Vec<_>>();
+    keys.extend(versions[1].keys());
+    keys.sort_unstable();
+    keys.dedup();
+
+    let _ = fs::remove_dir_all(SOURCE_DIR);
+    fs::rename(GENERATED_TEMP, SOURCE_DIR).unwrap();
+    assert!(versions[0] == versions[1]);
```

Comment:
I think it may be helpful to add some comments asking people to commit the changes. Just in case people updating other crates encounters this.

Reply:
Thanks for the review! I added a bunch of comments to make it clearer what's going on.